plutus-use-cases/src/Language/Plutus/Coordination/Contracts/Future.hs

```diff
@@ -0,0 +1,306 @@
+block). Likewise, if it moves against the buyer then the buyer has to top up
+their margin. If either party fails to make a margin payment then the contract
+will be settled early.
+
+The current value of the underlying asset is determined by an oracle. See note
+[Oracles] in Language.Plutus.Runtime.
+
+-}
+
+-- | Initialise the futures contract by paying the initial margin.
+--
+initialise :: (
+    MonadError WalletAPIError m,
+    WalletAPI m)
+    => PubKey
+    -- ^ Identity of the holder of the long position
+    -> PubKey
+    -- ^ Identity of the holder of the short position
+    -> Future
+    -> m ()
+initialise long short f = do
+    let
+        im = futureInitialMargin f
+        vl = fromIntegral im
+        o = scriptTxOut vl (validatorScript f) ds
+        ds = DataScript $ UTXO.lifted $ FutureData long short im im
+
+    (payment, change) <- createPaymentWithChange vl
+    signAndSubmit payment [o, change]
+
+-- | Close the position by extracting the payment
+settle :: (
+    MonadError WalletAPIError m,
+    WalletAPI m)
+    => [TxOutRef']
+    -> Future
+    -> FutureData
+    -> OracleValue Value
+    -> m ()
+settle refs ft fd ov = do
+    let
+        forwardPrice = futureUnitPrice ft
+        OracleValue (Signed (_, (_, spotPrice))) = ov
+        delta = futureUnits ft * (spotPrice - forwardPrice)
+        longOut = fromIntegral $ futureDataMarginLong fd + delta
+        shortOut = fromIntegral $ futureDataMarginShort fd - delta
+        red = UTXO.Redeemer $ UTXO.lifted $ Settle ov
+        outs = [
+            UTXO.pubKeyTxOut longOut (futureDataLong fd),
+            UTXO.pubKeyTxOut shortOut (futureDataShort fd)
+            ]
+        inp = (\r -> scriptTxIn r (validatorScript ft) red) <$> refs
+    signAndSubmit (Set.fromList inp) outs
+
+-- | Settle the position early if a margin payment has been missed.
+settleEarly :: (
+    MonadError WalletAPIError m,
+    WalletAPI m)
+    => [TxOutRef']
+    -> Future
+    -> FutureData
+    -> OracleValue Value
+    -> m ()
+settleEarly refs ft fd ov = do
+    let totalVal = fromIntegral $ futureDataMarginLong fd + futureDataMarginShort fd
+        outs = [UTXO.pubKeyTxOut totalVal (futureDataLong fd)]
+        inp = (\r -> scriptTxIn r (validatorScript ft) red) <$> refs
+        red = UTXO.Redeemer $ UTXO.lifted $ Settle ov
+    signAndSubmit (Set.fromList inp) outs
+
+adjustMargin :: (
+    MonadError WalletAPIError m,
+    WalletAPI m)
+    => [TxOutRef']
+    -> Future
+    -> FutureData
+    -> UTXO.Value
+    -> m ()
+adjustMargin refs ft fd vl = do
+    pk <- pubKey <$> myKeyPair
+    (payment, change) <- createPaymentWithChange vl
+    let
+        red = UTXO.Redeemer $ UTXO.lifted AdjustMargin
+        fd' = if pk == futureDataLong fd
+              then fd { futureDataMarginLong  = fromIntegral vl + futureDataMarginLong fd  }
+              else fd { futureDataMarginShort = fromIntegral vl + futureDataMarginShort fd }
```

Comment:
Should probably check it's the other key in this case and blow up if not. Not that someone else is going to adjust your margin for you.

Reply:
good point!